### PR TITLE
Analyzer support QUALIFY clause

### LINKED ISSUE #47819: QUALIFY Clause (Just like HAVING but for window functions)
**Use case**

```
SELECT <column_list>
  FROM <data_source>
  [GROUP BY ...]
  [HAVING ...]
  QUALIFY <predicate>
  [ ... ]


    WITH dense_rank() OVER (ORDER BY number DESC) AS row
    SELECT row
    FROM numbers(10)
    QUALIFY row < 3
```

**Describe alternatives you've considered**

Subquery

```
SELECT *
FROM
(
    WITH dense_rank() OVER (ORDER BY number DESC) AS row
    SELECT row
    FROM numbers(10)
)
WHERE row < 3
```

**Additional context**
https://github.com/ClickHouse/ClickHouse/issues/47818

https://docs.teradata.com/r/2_MC9vCtAJRlKle2Rpb0mA/19NnI91neorAi7LX6SJXBw
https://docs.snowflake.com/en/sql-reference/constructs/qualify

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Analyzer support QUALIFY clause. Closes #47819.

### CODE REVIEW

**Comment by kitaisreal** [comment]
More information https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#qualify_clause.

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit da061fba35aa853a3965e1942f69b468de421f4d with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62619/da061fba35aa853a3965e1942f69b468de421f4d/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: docs/en/sql-reference/statements/select/qualify.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+# QUALIFY Clause
+
+Allows filtering window functions results. It is similar to the [WHERE](../../../sql-reference/statements/select/where.md) clause, but the difference is that `WHERE` is performed before window functions evaluation, while `QUALIFY` is performed after it.
+
+It is possible to reference window functions results from `SELECT` clause in `QUALIFY` clause by their alias. Alternatively, `QUALIFY` clause can filter on results of additional window functions that are not returned in query results.
+
+## Limitations
+
+`QUALIFY` can’t be used if there are no window functions to evaluate. Use `WHERE` instead.
```
Comment (novikd) [inline docs/en/sql-reference/statements/select/qualify.md:14]:
```suggestion
`QUALIFY` can’t be used if there are no window functions to evaluate. Use `WHERE` or `HAVING` instead.
```

Comment (novikd) [reply docs/en/sql-reference/statements/select/qualify.md:14]: Let's also add an example to the docs?

File: tests/queries/0_stateless/03095_window_functions_qualify.sql [hunk 2]
```diff
@@ -0,0 +1,19 @@
+SET allow_experimental_analyzer = 1;
+
+SELECT number, COUNT() OVER (PARTITION BY number % 3) AS partition_count FROM numbers(10) QUALIFY partition_count = 4 ORDER BY number;
```
Comment (novikd) [inline tests/queries/0_stateless/03095_window_functions_qualify.sql:3]: Let's also add a test with a query like:
```sql
SELECT
  number,
  COUNT() OVER (PARTITION BY number % 3) AS partition_count
FROM numbers(10)
QUALIFY COUNT() OVER (PARTITION BY number % 3) = 4
ORDER BY number;
```
and check that window function is computed only once.

**Review by novikd — APPROVED** [review]
LGTM, but I'd like to see examples in the documentation and more complex tests.

File: tests/queries/0_stateless/03095_window_functions_qualify.sql (hunk 2 above)
Comment (kitaisreal) [reply tests/queries/0_stateless/03095_window_functions_qualify.sql:3]: Updated.

**Comment by clickgapai** [comment]
Hi @kitaisreal @alexey-milovidov @novikd — while reviewing this PR I found the following:

- **Coverage gap:** `mergeConditionNodes(HAVING, QUALIFY)` correctness untested — existing test uses identical conditions — https://github.com/ClickHouse/ClickHouse/pull/104073

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**